src/test-provider/analyze-test-run.ts

```diff
@@ -0,0 +1,43 @@
+import * as vscode from "vscode";
+import {
+  Failed,
+  SingleTestResult,
+  TestRunActions,
+  TestSuiteSummary,
+} from "./types";
+import { ansicolor } from "ansicolor";
+
+function failedGuard(result: SingleTestResult): result is Failed {
```

Comment:
I am not familiar with this syntax, what does it do?

Reply:
It's a filter function that also narrows the type of the kept elements, `result is Failed` is a `type predicate` and can be treated as a special form of a boolean. [See more about type rpedicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates)